src/sidebar/components/menu.js

```diff
@@ -0,0 +1,165 @@
+'use strict';
+
+const { Fragment, createElement } = require('preact');
+const { useCallback, useEffect, useRef, useState } = require('preact/hooks');
+const propTypes = require('prop-types');
+
+const { listen } = require('../util/dom');
+
+const SvgIcon = require('./svg-icon');
+
+let ignoreNextClick = false;
+
+/**
+ * A drop-down menu.
+ *
+ * Menus consist of a button which toggles whether the menu is open, an
+ * an arrow indicating the state of the menu and content when is shown when
+ * the menu is open. The children of the menu component are rendered as the
+ * content of the menu when open. Typically this consists of a list of
+ * `MenuSection` and/or `MenuItem` components.
+ *
+ * @example
+ *   <Menu label="Preferences">
+ *     <MenuItem label="View" onClick={showViewSettings}/>
+ *     <MenuItem label="Theme" onClick={showThemeSettings}/>
+ *   </Menu>
+ */
+function Menu({
+  align = 'left',
+  children,
+  defaultOpen = false,
+  label,
+  menuIndicator = true,
+  title,
+}) {
+  const [isOpen, setOpen] = useState(defaultOpen);
+
+  // Toggle menu when user presses toggle button. The menu is shown on mouse
+  // press for a more responsive/native feel but also handles a click event for
+  // activation via other input methods.
+  const toggleMenu = event => {
+    // If the menu was opened on press, don't close it again on the subsequent
+    // mouse up ("click") event.
+    if (event.type === 'mousedown') {
+      ignoreNextClick = true;
+    } else if (event.type === 'click' && ignoreNextClick) {
+      ignoreNextClick = false;
+      event.stopPropagation();
+      event.preventDefault();
+      return;
+    }
+    setOpen(!isOpen);
+  };
+  const closeMenu = useCallback(() => setOpen(false), [setOpen]);
+
+  // Close menu when user clicks outside or presses Esc.
+  const menuRef = useRef();
+  useEffect(() => {
```

Comment:
Let me walk through this because I've been staring at this block of code for a while and I want to see if I understand the lifecycle. Every time rendering happens, React will call the function passed to `useEffect`, which in this case is an anonymous function. Inside the function, if the menu is open, event listeners are bound to `document.body`. The anonymous function returns a function. Which could be fine...but I _cannot figure out_ how a return value from this function is used/invoked!

Reply:
UPDATED: I FINALLY GOT THIS! It took me half an hour. Please consider linking to this bit of docs in the comments because I nearly lost my mind here: https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect

> Often, effects create resources that need to be cleaned up before the component leaves the screen, such as a subscription or timer ID. To do this, the function passed to useEffect may return a clean-up function

is the bit that is relevant. So: that's how the returned function is invoked: as a cleanup.

I'm leaving all of these comments extant because it may be helpful to see my thinking process through this...some of this is on me as when I first read the docs for the effect hook, I didn't read them thoroughly enough, as it turns out.